analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotMarkEnumsWithFlags.cs

```diff
@@ -67,35 +68,49 @@ public sealed class DoNotMarkEnumsWithFlags : SonarDiagnosticAnalyzer
                     }
                 }, SyntaxKind.EnumDeclaration);
 
-        // The idea of this method is to get rid of invalid values for flags such as negative values and decimals
-        private static ulong? GetEnumValueOrDefault(IFieldSymbol enumMember) =>
-            enumMember.HasConstantValue && ulong.TryParse(enumMember.ConstantValue.ToString(), out var longValue)
-                ? longValue
+        private static BigInteger? GetEnumValueOrDefault(IFieldSymbol enumMember) =>
+            enumMember.HasConstantValue
+                ? enumMember.ConstantValue switch // BigInteger is used here to support enums with base type of ulong. Direct cast is used to avoid string conversion.
+                  {
+                      byte x => (BigInteger)x,
+                      sbyte x => (BigInteger)x,
+                      short x => (BigInteger)x,
+                      ushort x => (BigInteger)x,
+                      int x => (BigInteger)x,
+                      uint x => (BigInteger)x,
+                      long x => (BigInteger)x,
+                      ulong x => (BigInteger)x,
+                      // this is a safeguard for the case when the enum is defined with a base type that is not currently supported by the compiler
+                      var x when BigInteger.TryParse(x.ToString(), out var value) => value,
+                      _ => null
+                  }
                 : null;
 
-        private static bool IsValidFlagValue(ulong? enumValue, List<ulong> allValues) =>
+        private static bool IsValidFlagValue(BigInteger? enumValue, IEnumerable<BigInteger> allValues) =>
             enumValue.HasValue && (IsZeroOrPowerOfTwo(enumValue.Value) || IsCombinationOfOtherValues(enumValue.Value, allValues));
 
-        // See https://stackoverflow.com/questions/600293/how-to-check-if-a-number-is-a-power-of-2
-        private static bool IsZeroOrPowerOfTwo(ulong value) =>
-            (value & (value - 1)) == 0;
+        private static bool IsZeroOrPowerOfTwo(BigInteger value) =>
+            value.IsZero
+            || (value.Sign == -1
+                    ? BigInteger.Multiply(value, -1).IsPowerOfTwo
+                    : value.IsPowerOfTwo);
```

Comment:
BigInteger.Abs(value).IsPowerOfTwo
```suggestion
|| BigInteger.Abs(value).IsPowerOfTwo
```